### PR TITLE
feat(sse): ability to reconnect with new headers

### LINKED ISSUE #6046: feat: unstable_httpSubscriptionLink with headers polyfill does not update headers when reestablishing a lost connection
### Provide environment information

n/a

### Describe the bug

This is a known issue with event-source-polyfill https://github.com/Yaffle/EventSource/issues/137 but something we can probably fix in tRPC-land

Essentially EventSource is given headers and those headers are passed at instantiation. When a connection is lost EventSource will opaquely recover and reconnect, however it will use the original headers which may be outdated. This is a problem when the Authorization header contains a token which has expired since the EventSource was created

### Link to reproduction

https://github.com/trpc/trpc/pull/6047

### To reproduce

Pass a short lived header to unstable_httpSubscriptionLink.eventSourceOptions when using the polyfill, and then reboot the backend server. The header will be re-sent when connection is reestablished

### Additional information

_No response_

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/6046">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/6046/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by Nick-Lucas**
Closes #6046 

## 🎯 Changes

What changes are made in this PR? Is it a feature or a bug fix?

## ✅ Checklist

- [ ] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: packages/client/src/links/httpSubscriptionLink.ts [hunk 1]
```diff
@@ -140,3 +167,68 @@ export function unstable_httpSubscriptionLink<
...
+  private es: EventSource;
+
+  private listeners: Partial<Record<keyof EventSourceEventMap, any[][]>> = {};
+
+  constructor(url: string, options: EventSourceInit | undefined) {
+    this.es = new EventSource(url, options);
+  }
+
+  restart(url: string, options: EventSourceInit | undefined) {
+    this.es.close();
```
Comment (KATT) [inline packages/client/src/links/httpSubscriptionLink.ts:184]: you probably wanna prevent this from emitting any events

Comment (Nick-Lucas) [reply packages/client/src/links/httpSubscriptionLink.ts:191]: Agreed, haven't seen any negative effects, but no sense in risking it

Comment (Nick-Lucas) [reply packages/client/src/links/httpSubscriptionLink.ts:191]: This one is done, removed all known listeners before closing the connection

File: packages/client/src/links/httpSubscriptionLink.ts [hunk 2]
```diff
@@ -79,22 +88,22 @@ export function unstable_httpSubscriptionLink<
...
+                eventSource: eventSource!.getEventSource(),
               },
             });
-
-            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-            eventSource!.removeEventListener('open', onStarted);
           };
-          // console.log('starting', new Date());
-          eventSource.addEventListener('open', onStarted);
+          eventSource.addEventListener('open', onStarted, { once: true });
```
Comment (Nick-Lucas) [inline packages/client/src/links/httpSubscriptionLink.ts:105]: Important spec thing given the change in how this is run "once". When we do reinitialize, this callback will be called again with the new EventSource instance.

That felt right in-case anybody depends on holding that instance so they can update the reference and re-attach their own listeners

File: packages/client/src/links/httpSubscriptionLink.ts [hunk 3]
```diff
@@ -103,6 +112,24 @@ export function unstable_httpSubscriptionLink<
           >({
             from: eventSource,
             deserialize: transformer.output.deserialize,
+            tryHandleError: async (ev) => {
+              if (
```
Comment (Nick-Lucas) [inline packages/client/src/links/httpSubscriptionLink.ts:116]: we don't offer the chance to reboot when the network error itself occurs and that event doesn't come with a status code, but when it recovers and there's an issue like 401 we do call this. So it can take an extra round trip to recover but feels easier to work with

File: www/docs/client/links/httpSubscriptionLink.md [hunk 4]
```diff
@@ -178,8 +178,10 @@ If the client and server are not on the same domain, you can use `withCredential
 // [...]
 unstable_httpSubscriptionLink({
   url: 'https://example.com/api/trpc',
-  eventSourceOptions: {
-    withCredentials: true, // <---
+  eventSourceOptions() {
```
Comment (Nick-Lucas) [inline www/docs/client/links/httpSubscriptionLink.md:181]: I've removed the object variant of this because that way there is 0 nuance in when you can use it. Using a callback is a very small breaking change and gives tRPC more flexibility in calling it for newer values later

Comment (KATT) [reply www/docs/client/links/httpSubscriptionLink.md:181]: I don't see the motivation for it - what's the problem with `resultOf(opts.eventSourceOptions)`? 👀

File: packages/client/src/links/httpSubscriptionLink.ts [hunk 5]
```diff
@@ -37,7 +36,17 @@ type HTTPSubscriptionLinkOptions<TRoot extends AnyClientTypes> = {
...
+  /**
+   * For a given error, should we reinitialize the underlying EventSource?
+   *
+   * This is useful where a long running subscription might be interrupted by a recoverable network error,
+   * but the existing authorization header might have expired in the mean-time
+   */
+  shouldReinitialize?: (
+    status: number,
+    error: Event,
+  ) => boolean | Promise<boolean>;
```
Comment (KATT) [inline packages/client/src/links/httpSubscriptionLink.ts:49]: What if there's a transient error in the `eventSourceOptions`-callback? Should that follow the same code path?

File: packages/client/src/links/httpSubscriptionLink.ts (hunk 3 above)
Comment (KATT) [reply packages/client/src/links/httpSubscriptionLink.ts:116]: 🤔 

This feels a bit off to me but I don't hate it either... 

I feel like `shouldReinitialize()` should be called whenever it's closed while there is an active subscription

Comment (KATT) [reply packages/client/src/links/httpSubscriptionLink.ts:116]: Sketching

```ts
// closed unexpectedly 
type Closed = { type: "closed" }
// errors from the original EventSource
type Failure = { type: "event", event: Event; }

// if it's an error in i.e. the URL callback or the eventSourceOptions-callback 
type CallbackError = { type: "error", error: TRPCClientError }

type Failure = CallbackError | Closed | Failure
type ShouldReinitialize = (opts: Failure) => MaybePromise<boolean>;

```

File: packages/client/src/links/httpSubscriptionLink.ts [hunk 6]
```diff
@@ -37,7 +36,17 @@ type HTTPSubscriptionLinkOptions<TRoot extends AnyClientTypes> = {
...
    */
-  eventSourceOptions?: CallbackOrValue<EventSourceInit>;
+  eventSourceOptions?: () => EventSourceInit | Promise<EventSourceInit>;
+  /**
+   * For a given error, should we reinitialize the underlying EventSource?
+   *
+   * This is useful where a long running subscription might be interrupted by a recoverable network error,
+   * but the existing authorization header might have expired in the mean-time
+   */
+  shouldReinitialize?: (
```
Comment (KATT) [inline packages/client/src/links/httpSubscriptionLink.ts:46]: naming is hard, just thinking of alternatives

```suggestion
  shouldRecreate?: (
```

```suggestion
  shouldRecreateOnError?: (
```

**Review by KATT — COMMENTED** [review]
I like it but I think:

- I don't see why we need to remove the `eventSourceOptions`-object - I think the most common usecase is `{ withCredentials: true }`
- I think we might wanna make the `shouldReinitialize` more generic

let's discuss on Discord

File: packages/client/src/links/httpSubscriptionLink.ts (hunk 6 above)
Comment (KATT) [reply packages/client/src/links/httpSubscriptionLink.ts:46]: Maybe `shouldRecreateOn4xx` even if we want it to be specific

Currently naming suggests it's generic but it's very specific

File: packages/client/src/links/httpSubscriptionLink.ts (hunk 3 above)
Comment (Nick-Lucas) [reply packages/client/src/links/httpSubscriptionLink.ts:116]: I like the idea of that, I would also add a dedicated HttpError to that with status etc expanded because Event lacks the type safety here, but exposing the other states does make sense

File: www/docs/client/links/httpSubscriptionLink.md (hunk 4 above)
Comment (Nick-Lucas) [reply www/docs/client/links/httpSubscriptionLink.md:181]: We're still using resultOf internally

If the goal of shouldReinitialize is to get the latest copy of eventSourceOptions, then passing an object implicitly makes that functionality useless. Only allowing a callback means we don't need a load of conditional docs "use x if you're doing y, unless z then use f" becomes "use x if you want it, y if you need it"

File: packages/client/src/links/httpSubscriptionLink.ts (hunk 6 above)
Comment (Nick-Lucas) [reply packages/client/src/links/httpSubscriptionLink.ts:46]: `shouldRecreateOnError` I definitely prefer, think we can commit to that

File: packages/client/src/links/httpSubscriptionLink.ts (hunk 5 above)
Comment (Nick-Lucas) [reply packages/client/src/links/httpSubscriptionLink.ts:49]: I'd see that as a userland problem and we just let it bubble

File: www/docs/client/links/httpSubscriptionLink.md (hunk 4 above)
Comment (Nick-Lucas) [reply www/docs/client/links/httpSubscriptionLink.md:181]: Alternative would be a warning that if shouldReinitializeOnError is set, tRPC moans that eventSourceOptions needs to be a callback?

File: packages/client/src/links/httpSubscriptionLink.ts (hunk 3 above)
Comment (Nick-Lucas) [reply packages/client/src/links/httpSubscriptionLink.ts:127]: Right so I checked if there was any "Closed" type error via the test and doesn't seem to be, it may well fire on network disconnection but I've wrapped at "raw" and "http-error" types because I don't want to make too many assumptions about what non-http-error cases will look like but pulling out the status automatically makes a lot of sense as it's a common case

File: packages/client/src/links/httpSubscriptionLink.ts [hunk 7]
```diff
@@ -140,3 +203,74 @@ export function unstable_httpSubscriptionLink<
...
+
+  private listeners: Partial<Record<keyof EventSourceEventMap, any[][]>> = {};
+
+  constructor(url: string, options: EventSourceInit | undefined) {
+    this.es = new EventSource(url, options);
+  }
+
+  restart(url: string, options: EventSourceInit | undefined) {
+    for (const type in this.listeners) {
+      for (const [t, l, o] of this.listeners[
```
Comment (KATT) [inline packages/client/src/links/httpSubscriptionLink.ts:221]: - nested loop is usually bad
- don't do one character variable

File: packages/client/src/links/httpSubscriptionLink.ts [hunk 8]
```diff
@@ -140,3 +203,74 @@ export function unstable_httpSubscriptionLink<
...
   };
 }
+
+/**
+ * We wrap EventSource so that is can be reinitialized with new options
+ */
+class EventSourceWrapper implements Partial<EventSource> {
+  private es: EventSource;
+
+  private listeners: Partial<Record<keyof EventSourceEventMap, any[][]>> = {};
```
Comment (KATT) [inline packages/client/src/links/httpSubscriptionLink.ts:213]: looks like this could just be

```ts
  private listeners: Array<{
    type: keyof EventSourceEventMap;
    callback: (...args: any[]) => void
    options?: x
  }>
```

File: packages/client/src/links/httpSubscriptionLink.ts [hunk 9]
```diff
@@ -140,3 +203,74 @@ export function unstable_httpSubscriptionLink<
...
+    type: TEvent,
+    listener: (this: EventSource, ev: EventSourceEventMap[TEvent]) => any,
+    options?: boolean | EventListenerOptions,
+  ) {
+    this.listeners[type] ??= [];
+    for (const [t, l, o] of this.listeners[type] ?? []) {
+      if (t === type && l === listener) {
+        this.listeners[type].splice(this.listeners[type].indexOf([t, l, o]), 1);
+      }
+    }
```
Comment (KATT) [inline packages/client/src/links/httpSubscriptionLink.ts:272]: this can be simplified to something like

```ts
const idx = this.listeners.findIndex(x => /* ... */)
const listener = this.listeners[idx]
if (!listener) {
  return
}
this.es.removeEventListener(listener.type, listener.callback, listener.options)

// ... splice
```

File: packages/client/src/links/httpSubscriptionLink.ts [hunk 10]
```diff
@@ -103,6 +123,49 @@ export function unstable_httpSubscriptionLink<
...
+                'status' in ev &&
+                typeof ev.status === 'number' &&
+                typeof opts.shouldRecreateOnError === 'function' &&
+                eventSource
+              ) {
+                let recreateOnErrorOpts: RecreateOnErrorOpt = {
+                  type: 'raw',
+                  event: ev,
+                };
+                if ('status' in ev && typeof ev.status === 'number') {
```
Comment (KATT) [inline packages/client/src/links/httpSubscriptionLink.ts:137]: Avoid nested `if`s like this, this code is very hard to follow

do an early return or something instead

**Review by KATT — CHANGES_REQUESTED** [review]
I'm fine with the functionality and the interface but the code needs to be simplified - it's very hard to follow:

- Nested ifs
- Nested for loops
- One-character variable names

File: packages/client/src/links/httpSubscriptionLink.ts [hunk 11]
```diff
@@ -103,6 +123,49 @@ export function unstable_httpSubscriptionLink<
...
+                  return false;
+                }
+
+                if (typeof opts.eventSourceOptions !== 'function') {
+                  // eslint-disable-next-line no-console
+                  console.warn(
+                    'shouldRecreateOnError has returned true but eventSourceOptions is not a function. This means eventSourceOptions will not be updated',
+                  );
+                }
+
```
Comment (KATT) [inline packages/client/src/links/httpSubscriptionLink.ts:158]: Potential race condition here -

`shouldRestart` can be called twice before it gets a chance to restart

can you make the code in a way to avoid race conditions?

File: packages/client/src/links/httpSubscriptionLink.ts (hunk 8 above)
Comment (Nick-Lucas) [reply packages/client/src/links/httpSubscriptionLink.ts]: Type was slightly wrong with the `any` but best to use the Record as there are known keys and that makes adding and removing event listeners more efficient later

File: packages/client/src/links/httpSubscriptionLink.ts [hunk 12]
```diff
@@ -103,6 +123,49 @@ export function unstable_httpSubscriptionLink<
...
+                eventSource.restart(
+                  url,
```
[the remark this answers is not in the document]
Comment (Nick-Lucas) [reply packages/client/src/links/httpSubscriptionLink.ts]: Didn't realise that, I'll also remove the warning then since there's more than 1 reason to use this method

File: packages/client/src/links/httpSubscriptionLink.ts [hunk 13]
```diff
@@ -140,3 +192,97 @@ export function unstable_httpSubscriptionLink<
...
+  private listeners: Partial<
+    Record<
+      keyof EventSourceEventMap,
+      Parameters<EventSource['addEventListener']>[]
+    >
+  > = {};
+  private *getAllEventListeners() {
+    for (const _type in this.listeners) {
+      const type = _type as keyof EventSourceEventMap;
+      for (const listener of this.listeners[type] ?? []) {
```
Comment (KATT) [inline packages/client/src/links/httpSubscriptionLink.ts:226]: i'm guessing it could be
```suggestion
      const type = _type as keyof typeof this.listeners;
      for (const listener of this.listeners[type]) {
```

File: packages/server/src/unstable-core-do-not-import/stream/sse.ts [hunk 14]
```diff
@@ -227,19 +228,38 @@ export function sseStreamConsumer<TData>(opts: {
...
-  eventSource.addEventListener('error', (cause) => {
+  eventSource.addEventListener('error', async (cause) => {
+    // We prevent more than 1 error handler from waiting at the same time
+    const result = await errorLock;
+    if (result === 'CANCEL_ALL') {
+      return;
+    }
+
+    const resolvable = createResolvable<void | 'CANCEL_ALL'>();
+    errorLock = resolvable.promise;
```
Comment (KATT) [inline packages/server/src/unstable-core-do-not-import/stream/sse.ts:246]: we have a utility for this already called `createDeferred`
```suggestion
    const deferred = createDeferred<void | 'CANCEL_ALL'>();
    errorLock = deferred.promise;
```

File: packages/server/src/unstable-core-do-not-import/stream/sse.ts [hunk 15]
```diff
@@ -271,6 +291,20 @@ export function sseStreamConsumer<TData>(opts: {
...
+    resolve = _resolve;
+  });
+
+  if (resolve === NOOP) {
+    throw 'Unlock not assigned';
+  }
+
+  return { promise, resolve };
+}
+
```
Comment (KATT) [inline packages/server/src/unstable-core-do-not-import/stream/sse.ts:307]:
```suggestion
```

File: www/docs/client/links/httpSubscriptionLink.md [hunk 16]
```diff
@@ -113,6 +115,12 @@ const trpc = createTRPCClient<AppRouter>({
             },
           }; // you either need to typecast to `EventSourceInit` or use `as any` or override the types by a `declare global` statement
         },
+
+        // Optional: EventSource will automatically recover from disconnections, but always uses the initial headers
+        // If your auth header expires, you can force a restart and fresh eventSourceOptions() call when shouldRecreateOnError() returns true
+        shouldRecreateOnError(status, _error) {
+          return opts.type === 'http-error' && [401, 403].includes(opts.status);
+        },
```
Comment (KATT) [inline www/docs/client/links/httpSubscriptionLink.md:123]: I think this is worth its own header in the docs as well

Comment (Nick-Lucas) [reply www/docs/client/links/httpSubscriptionLink.md]: Fair point, done and cleaned up this one since it becomes just noise

File: packages/client/src/links/httpSubscriptionLink.ts (hunk 11 above)
Comment (Nick-Lucas) [reply packages/client/src/links/httpSubscriptionLink.ts]: Above comment is outdated

Approach now just locks the error stream, and if an error gets handled will discard the waiting stream to avoid restarting again immediately. Feels like only compromises in this space so simple is good as a first approach I think

**Review by KATT — APPROVED** [review]
